File: dsa/week(4)/Order_con_des.cpp
// Vehicle (base class)  <------- Car (derived class of Vehicle) <---------  HondaCity (derived class of Car)
// Vehicle Constructor <--- Car Constructor <--- HondaCity Constructor <--- HondaCity Destructor <---  Car Destructor <--  Vehicle Destructor

#include <iostream>
using namespace std;

class Vehicle {
private:
    int maxSpeed;
protected:
    int numTyres;
public:
    string color;

    Vehicle(int z) {
        cout << "Vehicle's Parametrized constructor" << endl;
        maxSpeed = z;
    }
    ~Vehicle() {
        cout << "Vehicle's Destructor" << endl;
    }
};

class Car : public Vehicle {
public:
    int numGears;
    Car(int x, int y) : Vehicle(x) {
        cout << "Car's constructor" << endl;
        numGears = y;
    }
    ~Car() {
        cout << "Car's Destructor" << endl;
    }

    void print() {
        cout << "NumTyres: " << numTyres << endl;
        cout << "Color: " << color << endl;
        cout << "Num gears: " << numGears << endl;
        
    }
};

class HondaCity : public Car {
public:
    HondaCity(int x, int y) : Car(x, y) {
        cout << "Honda city constructor" << endl;
    }

    ~HondaCity() {
        cout << "Honda city destructor" << endl;
    }
};

int main() {
    HondaCity h(4, 5);
    return 0;
}

File: dsa/week(4)/Order_con_des.py
class Vehicle:
    def __init__(self, z):
        print("Vehicle's Parametrized constructor")
        self.__maxSpeed = z 
        self._numTyres = None  
        self.color = None

    def __del__(self):
        print("Vehicle's Destructor")


class Car(Vehicle):
    def __init__(self, x, y):
        super().__init__(x) 
        print("Car's constructor")
        self.numGears = y

    def __del__(self):
        print("Car's Destructor")
        super().__del__()

    def print_details(self):
        print("NumTyres:", self.numTyres)
        print("Color:", self.color)
        print("Num gears:", self.numGears)
       


class HondaCity(Car):
    def __init__(self, x, y):
        super().__init__(x, y) 
        print("Honda city constructor")

    def __del__(self):
        print("Honda city destructor")
        super().__del__()





h = HondaCity(4, 5)
del h  # Explicitly deleting to simulate destructor
